pybart/new_matcher.py

```diff
@@ -0,0 +1,330 @@
+                    pattern["ENT_TYPE"] = {"IN", spec.value}
+            
+            patterns.append((constraint.id, not constraint.optional, pattern))
+        return patterns
+    
+    def __init__(self, constraints: Sequence[Token]):
+        self.nlp = spacy.load("en_ud_model_lg")
+        self.matcher = SpacyMatcher(self.nlp.vocab)
+        # store the incoming/outgoing token constraints according to the token id for post token-level matching
+        self.incoming_constraints = {constraint.id: constraint.incoming_edges for constraint in constraints}
+        self.outgoing_constraints = {constraint.id: constraint.outgoing_edges for constraint in constraints}
+        
+        # convert the constraint to a list of matchable token patterns
+        self.required_tokens = set()
+        for token_name, is_required, matchable_pattern in self._convert_to_matchable(constraints):
+            # add it to the matchable list
+            self.matcher.add(token_name, None, matchable_pattern)
+            if is_required:
+                self.required_tokens.add(token_name)
+    
+    def _post_spacy_matcher(self, matched_tokens: Mapping[str, List[int]], sentence) -> Mapping[str, List[int]]:
+        # handles incoming and outgoing label constraints (still in token level)
+        checked_tokens = dict()
+        for name, token_indices in matched_tokens.items():
+            checked_tokens[name] = [token for token in token_indices if
+                                    are_labels_satisfied(self.incoming_constraints[name], sentence.get_labels(child=token))[0] and
+                                    are_labels_satisfied(self.outgoing_constraints[name], sentence.get_labels(parent=token))[0]]
+        return checked_tokens
+    
+    def apply(self, sentence) -> Mapping[str, List[int]]:  # TODO - define the Graph class we will work with as sentence
+        matched_tokens = defaultdict(list)
+        
+        # apply spacy's token-level match
+        matches = self.matcher(sentence.doc)
+        # validate the span and store each matched token
+        for match_id, start, end in matches:
+            token_name = self.nlp.vocab.strings[match_id]
+            # assert that we matched no more than single token
+            if end - 1 == start:
+                raise ValueError("matched more than single token")  # TODO - change to our Exception for handling more gently
+            token = sentence.doc[start]
+            matched_tokens[token_name].append(token.id)
+        
+        # reverse validate the 'optional' constraint
+        if len(self.required_tokens.difference(set(matched_tokens.keys()))) != 0:
+            raise ValueError("required token not matched")  # TODO - change to our Exception for handling more gently
+        
+        # extra token matching out of spacy's scope
+        matched_tokens = self._post_spacy_matcher(matched_tokens, sentence)
+        
+        return matched_tokens
+
+
+class Matchers(NamedTuple):
```

Comment:
why do you need this?

Reply:
```python
def matches_for(self, name: str) -> Generator[MatchingResult, None, None]:
    # token match
    matches = self.matchers[name].token_matcher.apply(self.sentence)
    
    # filter
    yield from self.matchers[name].global_matcher.apply(matches, self.sentence)
```

each time I access access a matchers object by name I need to access the token_matcher and global_matcher separately and I think it is more readable than accessing a (non-named) tuple: `self.matchers[name][0].apply()`